add-okved-client-api

haribo27 opened the pull request:
Сделал api-client , который по номеру оквед получает раздел платежа. 
Сделал бизнес логику по сохранению платежа.
Создал скрипт создания таблицы schema.sql

haribo27 committed:
В пакете updater.messaging добавил функционал приема сообщений(json) от kafka producer и передача данных в сервис

haribo27 committed:
Устранил баги, добавил логирование в сервис.

haribo27 committed:
Устранил баги, добавил логирование в сервис.

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/model/Payment.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
```
В Entity лучше не делать примитивные атрибуты - их нужно сделать Long или BigDecimal

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/okvedApiClient/client/OkvedApiClient.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Value("${okved.api-key}")
+    private String apiKey;
+    private static final String OKVED_FIND_BY_ID_URL =
```
Этот url лучше в конфиги вынести

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/okvedApiClient/client/OkvedApiClient.java:
```diff
@@ -0,0 +1,39 @@
+        String query = "{\"query\": \"" + okvedId + "\"}";
+        HttpEntity<String> requestEntity = new HttpEntity<>(query, defaultHeaders());
+
+        return restTemplate.exchange(OKVED_FIND_BY_ID_URL,
```
Тут не хватает обработки ошибки, в том случае, если получим ошибку при вызове api - ее нужно, как минимум, логгировать. В общем, нужно подумать над обработкой исключительных ситуаций

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/okvedApiClient/client/OkvedApiClient.java:
```diff
@@ -0,0 +1,39 @@
+    private static final String OKVED_FIND_BY_ID_URL =
+            "http://suggestions.dadata.ru/suggestions/api/4_1/rs/findById/okved2";
+
+    private final RestTemplate restTemplate;
```
Не как замечание, а как пожелание на будущее - попробуй переделать на Feign Client - сейчас это более распространенный подход для вызовов REST API

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/okvedApiClient/service/OkvedService.java:
```diff
@@ -0,0 +1,32 @@
+        } catch (JsonProcessingException e) {
+            System.out.println(e.getMessage());
+        }
+        JsonNode suggestionsNode = root.path("suggestions");
```
Лучше ответ апишки сразу после его получения маппить в какой-нибудь ДТО, так с ним будет проще работать потом, без необходимости использования JSON-классов. И так мы уберем логику маппинга ответа из бизнесового бина

PavelAndryushenko commented on updater/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE IF NOT EXISTS payments (
```
Создание схемы и объектов БД лучше сделать при помощи миграций - Liquibase или Flyway

PavelAndryushenko commented on updater/src/test/java/ru/zubcov/paymentstats/updater/UpdaterApplicationTests.java:
```diff
@@ -0,0 +1,15 @@
+package ru.zubcov.paymentstats.updater;
```
Если этот тестовый класс не нужен, лучше его вовсе удалить

haribo27 committed:
Исправил замечания
Добавил версионирование базы данных flyway
Изменил клиент для апи запросов с restTemplate на feign client

haribo27 committed:
Небольшие правки в окведсервис, добавил выброс исключений в случае не корректного запроса или пустого ответа

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/model/Payment.java:
```diff
@@ -0,0 +1,33 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(name = "client_id", nullable = false)
+    private long clientId;
```
Остались еще примитивы в Entity - long, double

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/model/Payment.java:
```diff
@@ -0,0 +1,33 @@
+    private String clientAccount;
+    @Column(name = "amount", nullable = false)
+    @PositiveOrZero
+    private double amount;
```
В нашем проекте это не критично, но в коммерческих проектах лучше использовать BigDecimal, поскольку double все же имеет потерю точности

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/okvedApiClient/service/OkvedService.java:
```diff
@@ -0,0 +1,41 @@
+                throw new UnexpectedApiResponseError("Some of response fields = null");
+            }
+        } else {
+            log.info("Request status: {}", response.getStatusCode());
```
Тут явно мы имеем дело с какой-то ошибкой - лучше уровень логирования повысить до warn

PavelAndryushenko commented on updater/src/main/java/ru/zubcov/paymentstats/updater/okvedApiClient/dto/SuggestionResponseDto.java:
```diff
@@ -0,0 +1,42 @@
+
+import java.util.List;
+
+@Data
```
Это класс - DTO, его лучше сделать иммутабельным

PavelAndryushenko reviewed (approved):
В целом, можно мерджить, а замечания можем поправить в отдельной задаче, они не очень критичные

haribo27 committed:
Поправил замечания

haribo27 committed:
Merge branch 'master' into add-okved-client-api

# Conflicts:
#	pom.xml
#	stats/pom.xml
#	updater/pom.xml
#	updater/src/main/java/ru/zubcov/paymentstats/updater/messaging/payment/PaymentDto.java
#	updater/src/main/java/ru/zubcov/paymentstats/updater/messaging/service/KafkaMessagingService.java
#	updater/src/main/java/ru/zubcov/paymentstats/updater/repository/PaymentRepository.java
#	updater/src/main/java/ru/zubcov/paymentstats/updater/service/PaymentServiceImpl.java
#	updater/src/main/resources/application.properties

haribo27 committed:
Небольшие изменения в дто классе, чуть не верно смерджил
